### PR TITLE
[REEF-943] Decouple the DriverSide heartbeat handling from protobuf

### PR DESCRIPTION
**Pull request by zerg-junior**
This patch:
 * Adds the 'pojos' package to 'org.apache.reef.runtime.common.driver.evaluator'
 * Implements EvaluatorStatusPOJO, ContextStatusPOJO, TaskStatusPOJO, ContextMessagePOJO, TaskMessagePOJO.
   These classes turn ReefServiceProtos into corresponding POJOs. The POJO classes implement only the methods
   that Driver currently uses; POJOs discard all other proto fields.
 * Implements ContextState and State enums that mimic corresponding enums from ReefServiceProtos.
   In the original code with protos the State enum represents states of  Tasks, Evaluators and ResourceManagers.
   This patch adopts this policy.
 * Replaces ReefServiceProtos with corresponding POJOs and enums in all event handles that
   EvaluatorManager#onEvaluatorHeartbeatMessage calls

JIRA:
 [REEF-943] (https://issues.apache.org/jira/browse/REEF-943)

### CODE REVIEW

**Comment by markusweimer** [comment]
I'll do a pass.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.reef.annotations.audience.DriverSide;
+import org.apache.reef.annotations.audience.Private;
+import org.apache.reef.proto.ReefServiceProtos;
+
+/**
+ *  DriverSide representation of ContextMessageProto.
+ */
+@DriverSide
+@Private
+public class ContextMessagePOJO {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java]: `final`

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  private byte[] message;
+  private String sourceId;
+
+  public ContextMessagePOJO(final ReefServiceProtos.ContextStatusProto.ContextMessageProto proto){
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java]: Does this constructor need to be `public`? The same question arises for all the other newly added classes / constructors. I did some checks and for instance the one for `TaskMessage` can be package-private. Please check and make the new constructors the least visible.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.reef.proto.ReefServiceProtos;
+
+/**
+ *  DriverSide representation of ContextMessageProto.
+ */
+@DriverSide
+@Private
+public class ContextMessagePOJO {
+
+  private byte[] message;
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java]: Please make all instance variables `final`. Where that is not possible, please document the thread safety concerns.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  private String sourceId;
+
+  public ContextMessagePOJO(final ReefServiceProtos.ContextStatusProto.ContextMessageProto proto){
+    message = proto.getMessage().toByteArray();
+    sourceId = proto.getSourceId();
+  }
+
+  public byte[] getMessage(){
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextMessagePOJO.java]: Please document all new `public` methods.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextState.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.reef.annotations.audience.DriverSide;
+import org.apache.reef.annotations.audience.Private;
+
+/**
+ * DriverSide representation of ContextStatusProto.State.
+ */
+@DriverSide
+@Private
+public enum ContextState {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/ContextState.java]: `final`

**Comment by markusweimer** [comment]
I've done a pass and left some broad feedback. Let me know when that has been addressed and I will do another one.

**Comment by zerg-junior** [comment]
@markusweimer  Updated

**Comment by markusweimer** [comment]
I'm at a conference with very bad internet and won't be able to review and merge before I am back. So if anyone else can do that, please go ahead.

**Comment by markusweimer** [comment]
@zerg-junior Can you please resolve conflicts and rebase?

**Comment by zerg-junior** [comment]
@markusweimer Rebased. I can't locally reproduce the `FailDriverDelayedMsgTest`  failure.

**Pull request closed without merging**